servicemanager/servicemanager.go

```diff
@@ -0,0 +1,260 @@
+ * Public
+ *********************************************************************************************************************/
+
+// New creates new launcher object.
+func New(config *config.Config, serviceInfoProvider ServiceStorage) (sm *ServiceManager, err error) {
+	sm = &ServiceManager{
+		servicesDir:         config.ServicesDir,
+		downloadDir:         config.DownloadDir,
+		actionHandler:       action.New(maxConcurrentActions),
+		gidPool:             uidgidpool.NewGroupIDPool(),
+		serviceInfoProvider: serviceInfoProvider,
+	}
+
+	if err = os.MkdirAll(sm.servicesDir, 0o755); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	services, err := sm.serviceInfoProvider.GetAllServices()
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	for _, service := range services {
+		if err = sm.gidPool.AddID(service.GID); err != nil {
+			log.Errorf("Can't add service GID to pool: %s", err)
+		}
+	}
+
+	return sm, nil
+}
+
+// InstallService install service to the system.
+func (sm *ServiceManager) InstallService(
+	newService ServiceInfo, imageURL string, fileInfo image.FileInfo) (err error) {
+	return <-sm.actionHandler.Execute(newService.ServiceID,
+		func(id string) error {
+			return sm.doInstallService(newService, imageURL, fileInfo)
+		},
+	)
+}
+
+func (sm *ServiceManager) GetServiceInfo(serviceID string) (serviceInfo ServiceInfo, err error) {
+	serviceInfo, err = sm.serviceInfoProvider.GetService(serviceID)
+	return serviceInfo, aoserrors.Wrap(err)
+}
+
+func (sm *ServiceManager) GetImageParts(service ServiceInfo) (parts ImageParts, err error) {
+	return getImageParts(service.ImagePath)
+}
+
+/**********************************************************************************************************************
+ * Private
+ *********************************************************************************************************************/
+
+func (sm *ServiceManager) doInstallService(newService ServiceInfo, imageURL string, fileInfo image.FileInfo) error {
+	serviceFromStorage, err := sm.serviceInfoProvider.GetService(newService.ServiceID)
+	if err != nil && !errors.Is(err, ErrNotExist) {
+		return aoserrors.Wrap(err)
+	}
+
+	var isServiceExist bool
+
+	if err == nil {
+		isServiceExist = true
+
+		if newService.AosVersion <= serviceFromStorage.AosVersion {
+			return ErrVersionMismatch
+		}
+	}
+
+	// download and unpack
+	urlVal, err := url.Parse(imageURL)
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	var serviceImage string
+
+	if urlVal.Scheme != "file" {
+		if serviceImage, err = image.Download(context.Background(), sm.downloadDir, imageURL); err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		defer os.RemoveAll(serviceImage)
+	} else {
+		serviceImage = urlVal.Path
+	}
+
+	if err = image.CheckFileInfo(context.Background(), serviceImage, fileInfo); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	// create install dir
```

Comment:
Remove comment. Or move to appropriate place.

Reply:
removed